wasm/src/lib.rs

```diff
@@ -109,6 +123,53 @@ fn _get_qir(sources: SourceMap, language_features: LanguageFeatures) -> Result<S
     generate_qir(&store, package).map_err(|e| e.0.to_string())
 }
 
+fn _get_qir_preview(
+    sources: SourceMap,
+    language_features: LanguageFeatures,
+    profile: Profile,
+) -> Result<String, String> {
+    let core = compile::core();
+    let mut package_store = PackageStore::new(core);
+    let std = compile::std(&package_store, profile.into());
+    let std = package_store.insert(std);
+
+    let (unit, errors) = qsc::compile::compile(
+        &package_store,
+        &[std],
+        sources,
+        PackageType::Exe,
+        profile.into(),
+        language_features,
+    );
+
+    // Ensure it compiles before trying to add it to the store.
+    if !errors.is_empty() {
+        // This should never happen, as the program should be checked for errors before trying to
+        // generate code for it. But just in case, simply report the failure.
+        return Err("Failed to generate QIR".to_string());
+    }
+
+    let package_id = package_store.insert(unit);
+
+    let caps_results =
+        PassContext::run_fir_passes_on_hir(&package_store, package_id, profile.into());
+    // Ensure it compiles before trying to add it to the store.
+    match caps_results {
+        Ok(compute_properties) => hir_to_qir(
+            &package_store,
+            package_id,
+            profile.into(),
+            Some(compute_properties),
+        )
+        .map_err(|e| e.to_string()),
+        Err(_) => {
+            // This should never happen, as the program should be checked for errors before trying to
+            // generate code for it. But just in case, simply report the failure.
+            Err("Failed to generate QIR".to_string())
+        }
+    }
+}
```

Comment:
Again, too much behavior for what's supposed to be a thin interop layer (the wasm crate). Shouldn't all this code live in the `qsc_codegen` crate? Surely we'll want to share this code from Python too.

Reply:
This can't be in the codegen crate as it would create a circular dependency